jina/peapods/runtimes/zed.py

```diff
@@ -0,0 +1,226 @@
+import os
+import time
+from collections import defaultdict
+from typing import Dict, List
+
+import zmq
+
+from jina.peapods.runtimes import BaseRuntime
+from ..zmq import ZmqStreamlet, send_ctrl_message, Zmqlet
+from ... import Message
+from ... import Request
+from ...enums import SkipOnErrorType
+from ...excepts import NoExplicitMessage, ExecutorFailToLoad, MemoryOverHighWatermark, ChainedPodException, \
+    BadConfigSource, RuntimeTerminated
+from ...executors import BaseExecutor
+from ...logging.profile import used_memory, TimeDict
+from ...proto import jina_pb2
+
+
+class ZEDRuntime(BaseRuntime):
+    def run_forever(self):
+        self._zmqlet.start(self._msg_callback)
+
+    def cancel(self):
+        ctrl_addr, _ = Zmqlet.get_ctrl_address(self.args.host, self.args.port_ctrl, self.args.ctrl_with_ipc)
+        send_ctrl_message(ctrl_addr, 'TERMINATE', timeout=self.args.timeout_ctrl)
+
+    def setup(self):
+        self._last_active_time = time.perf_counter()
+        self._last_dump_time = time.perf_counter()
+
+        self._timer = TimeDict()
+
+        self._request = None
+        self._message = None
+
+        # all pending messages collected so far, key is the request id
+        self._pending_msgs = defaultdict(list)  # type: Dict[str, List['Message']]
+        self._partial_requests = None
+        self._partial_messages = None
+        self._zmqlet = ZmqStreamlet(self.args, logger=self.logger)
+
+        self._load_plugins()
+        self._load_executor()
+
+    def teardown(self):
+        self._zmqlet.close()
+        self._executor.close()
+
+    #: Private methods required by :meth:`setup`
+
+    def _load_executor(self):
+        """Load the executor to this BasePea, specified by ``uses`` CLI argument.
+        """
+        try:
+            try:
+                self._executor = BaseExecutor.load_config(self.args.uses,
```

Comment:
a big benefit of the latest refactor is to have executor context managed by the Pea which is a very robust design. Can we keep it this way?

Reply:
is there any plan to benefit of this? that would give a lot of robustness in ensuring the proper closure of Executors. Instead of trusting that the developer will safely capture every potential problem

Reply:
>  executor context managed , ... , is there any plan to benefit of this?

There is no magic on context manager. It relies on handwritten code such as `.start()`, `.close`. Previously one of the big headache is you never now when the context manager is ending its context, in main process/subprocess, in side forever loop or outside forever loop. It is unrealistic to just believe `with ...` will magically handle everything.

So it's not part of the PR, as I see it is a sugary syntax which only weakly benefits people who work on `Runtime` but no others. Further improvement in separate PR is welcome but please first understand all corner cases described in the `BasePea:start()`